#This patch file is from Qualcomm Innovation Center, Inc. and is provided under the following license:
#
#Copyright (c) 2025 Qualcomm Innovation Center, Inc. All rights reserved.
#SPDX-License-Identifier: BSD-3-Clause-Clea
From c8578da7c811956e2fa3e71f9b0077ca543665ba Mon Sep 17 00:00:00 2001
From: Venkatesh Kotha <quic_venkkoth@quicinc.com>
Date: Sat, 22 Mar 2025 19:36:52 +0530
Subject: DNM weston: temp fix for avoid compilation

Upstream-Status: Inappropriate [Downstream]

Signed-off-by: Venkatesh Kotha <quic_venkkoth@quicinc.com>
---
 .../backend-sdm/sdm-service/sdm_display.cpp   | 12 +++-
 .../sdm_display_buffer_allocator.cpp          | 57 +++++++++++++++++--
 2 files changed, 62 insertions(+), 7 deletions(-)

diff --git a/libweston/backend-sdm/sdm-service/sdm_display.cpp b/libweston/backend-sdm/sdm-service/sdm_display.cpp
index e70ef1de..e19c0bb5 100644
--- a/libweston/backend-sdm/sdm-service/sdm_display.cpp
+++ b/libweston/backend-sdm/sdm-service/sdm_display.cpp
@@ -72,7 +72,7 @@
 #include <pthread.h>
 #include <sys/prctl.h>
 #include <sys/resource.h>
-#include <gbm_msm.h>
+#include <gbm_priv.h>
 
 #include <gbm.h>
 
@@ -818,9 +818,15 @@ int SdmDisplay::PrepareNormalLayerGeometry(struct drm_output *output,
             uint32_t ubwc_status = 0;
             void *prm = reinterpret_cast<void *> (&layer->color_metadata);
 
-            gbm_msm_bo *msm_bo = reinterpret_cast<gbm_msm_bo*>(bo);
+	    gbm_perform(GBM_PERFORM_GET_BO_ALIGNED_WIDTH, bo, &alignedWidth);
+            gbm_perform(GBM_PERFORM_GET_BO_ALIGNED_HEIGHT, bo, &alignedHeight);
+            gbm_perform(GBM_PERFORM_GET_SECURE_BUFFER_STATUS, bo, &secure_status);
+            gbm_perform(GBM_PERFORM_GET_METADATA, bo, GBM_METADATA_GET_COLOR_METADATA, prm);
+            gbm_perform(GBM_PERFORM_GET_UBWC_STATUS, bo, &ubwc_status);
+
+/*	    gbm_msm_bo *msm_bo = reinterpret_cast<gbm_msm_bo*>(bo);
 	          alignedWidth = msm_bo->aligned_width;
-	          alignedHeight = msm_bo->aligned_height;
+	          alignedHeight = msm_bo->aligned_height; */
 
             // Override buffer width/height to reflect aligned width and aligned height.
             layer->width = alignedWidth;
diff --git a/libweston/backend-sdm/sdm-service/sdm_display_buffer_allocator.cpp b/libweston/backend-sdm/sdm-service/sdm_display_buffer_allocator.cpp
index 4431196d..be1c8a7a 100644
--- a/libweston/backend-sdm/sdm-service/sdm_display_buffer_allocator.cpp
+++ b/libweston/backend-sdm/sdm-service/sdm_display_buffer_allocator.cpp
@@ -30,7 +30,7 @@
 #include "sdm-service/sdm_display_debugger.h"
 #include "sdm-service/sdm_display_buffer_allocator.h"
 #include <gbm.h>
-#include <gbm_msm.h>
+#include <gbm_priv.h>
 
 #define __CLASS__ "SdmDisplayBufferAllocator"
 
@@ -101,7 +101,6 @@ int SdmDisplayBufferAllocator::AllocateBuffer(BufferInfo *buffer_info) {
   uint32_t width = buffer_config.width;
   uint32_t height = buffer_config.height;
   uint32_t format;
-  int metadata_fd = -1;
   uint64_t alloc_flags = 0;
   uint32_t ubwc_status;
   int error = SetBufferInfo(buffer_config.format, &format, &alloc_flags);
@@ -114,6 +113,20 @@ int SdmDisplayBufferAllocator::AllocateBuffer(BufferInfo *buffer_info) {
   if (bo) {
       alloc_buffer_info->fd = gbm_bo_get_fd(bo);
       alloc_buffer_info->stride = gbm_bo_get_stride(bo);
+
+      uint32_t alignedWidth = 0;
+      gbm_perform(GBM_PERFORM_GET_BO_ALIGNED_WIDTH, bo, &alignedWidth);
+      alloc_buffer_info->aligned_width = alignedWidth;
+      uint32_t alignedHeight = 0;
+      gbm_perform(GBM_PERFORM_GET_BO_ALIGNED_HEIGHT, bo, &alignedHeight);
+      alloc_buffer_info->aligned_height = alignedHeight;
+      size_t bo_size = 0;
+      gbm_perform(GBM_PERFORM_GET_BO_SIZE, bo, &bo_size);
+      gbm_perform(GBM_PERFORM_GET_UBWC_STATUS, bo, &ubwc_status);
+      alloc_buffer_info->size = (uint32_t)bo_size;
+
+#if 0
+      int metadata_fd = -1;
       uint32_t alignedWidth = 0;
       uint32_t alignedHeight = 0;
       gbm_msm_bo *msm_bo = reinterpret_cast<gbm_msm_bo*>(bo);
@@ -124,6 +137,7 @@ int SdmDisplayBufferAllocator::AllocateBuffer(BufferInfo *buffer_info) {
       uint32_t bo_size = 0;
       bo_size = msm_bo->size;
       alloc_buffer_info->size = bo_size;
+#endif
       alloc_buffer_info->format = GetLayerBufferFormat(gbm_bo_get_format(bo), ubwc_status);
   } else {
       DLOGE("Failed to allocate memory");
@@ -283,7 +297,7 @@ int SdmDisplayBufferAllocator::GetBufferLayout(const AllocatedBufferInfo &buf_in
     uint32_t format1 = GBM_FORMAT_ARGB8888;
     uint64_t flags = 0;
     uint32_t ubwc_status = 0;
-    uint32_t aligned_width = 0;
+    generic_buf_layout_t buf_layout;
 
     SetBufferInfo(buf_info.format, &format1, &flags);
 
@@ -303,8 +317,42 @@ int SdmDisplayBufferAllocator::GetBufferLayout(const AllocatedBufferInfo &buf_in
 
     uint32_t height = gbm_bo_get_height(bo);
     uint32_t format = gbm_bo_get_format(bo);
-    uint32_t bpp = gbm_bo_get_bpp(bo);
 
+    if (IsFormatVideo(format) == false) {
+      stride[0] = gbm_bo_get_stride(bo);
+      offset[0] = 0;
+      *num_planes++;
+    } else {
+      /*
+        TODO: A more elegant solution would be to create a GBM API call to
+              get number of non-meta planes.
+              Additionally, need to check YUV formats using a GBM API so as to keep
+              the implementation generic. Currently, this is a local function call limited
+              to 3 YUV formats, and every time a new format needs to be introduced,
+              implementation of IsFormatVideo needs to be updated.
+      */
+      *num_planes = 2;  // For video formats
+      gbm_perform(GBM_PERFORM_GET_PLANE_INFO, bo, &buf_layout);
+      gbm_perform(GBM_PERFORM_GET_UBWC_STATUS, bo, &ubwc_status);
+ 
+      if (IsVideoFormatLinear(format, ubwc_status)) {
+        stride[0] = buf_layout.planes[0].v_increment;
+        offset[0] = 0;
+ 
+        stride[1] = stride[0];
+        offset[1] = stride[0]*height;
+      }
+      if (IsVideoFormatUBWC(format, ubwc_status)) {
+        stride[0] = buf_layout.planes[0].v_increment;
+        stride[1] = buf_layout.planes[1].v_increment;
+        offset[0] = 0;
+        offset[1] = 0;
+      }
+    }
+
+#if 0
+    uint32_t aligned_width = 0;
+    uint32_t bpp = gbm_bo_get_bpp(bo);
     if (IsFormatVideo(format) == false) {
       /* Importing bo from mesa-gbm requires stride value to be passed as part of
       gbm_import_fd_data structure. Our gbm implementation doesn't mandate this requirement.
@@ -326,6 +374,7 @@ int SdmDisplayBufferAllocator::GetBufferLayout(const AllocatedBufferInfo &buf_in
 
     if (format == GBM_FORMAT_NV12)
       *num_planes = 2; // for NV12 format
+#endif
 
   gbm_bo_destroy(bo);
 
-- 
2.49.0

